### PR TITLE
feature: add release automation workflows

### PR DESCRIPTION
**Pull request by tuunit**
## Description
To stabilize and accelerate the frequency of releases a proper release automation is necessary. This PR introduces two new workflows 1. `Create Release` and 2. `Publish Release`.

1. Create Release
  - This workflow has to be triggered manually with a new version
  - Updates the documentation and CHANGELOG
  - Automatically creates a release branch e.g. `release/v7.5.1` and raises a PR for it
2. Publish Release
  - This workflow is automatically triggered when a PR is merged to the master branch but only runs through if it was merged from a release branch
  - Creates a pre-release on GitHub
  - Builds all necessary oauth2-proxy binaries and uploads them (including the checksums) to the GitHub pre-release
  - Builds all platform container images and publish them to quay.io

Only steps not automated:
- Updating the CHANGELOG with highlight and breaking changes
- Updating and publishing the Pre-Release on GitHub with the CHANGELOG content

For this automation to work the following has be setup in GitHub:
1. Allow pull request creation through github actions in the org or repo settings
2. Create a robo account for quay.io/oauth2-proxy and create a `REGISTRY_USERNAME` and `REGISTRY_PASSWORD` action secret for the repo.

## How Has This Been Tested?
To properly test this PR a completely separate repository has been created:
https://github.com/tuunit/oauth2-proxy-release-automation

## Checklist:

- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

**Comment by tuunit** [comment]
> This LGTM, although I would still appreciate to separate Releases of the oauth2-proxy from building docker image, as I think, that we need to build new docker images nightly, even if there are no code changes in this repo, since the base images need to be updated. But this can be done in a separate PR.

I would prefer nightly rebuilds of the latest one or two releases as well.

File: .github/workflows/create-release.yml [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        git checkout -b release/v${VERSION}
+        git commit -am "update to release version v${VERSION}"
+        git push -u origin release/v${VERSION}
+    
+    - name: Create PR
+      env:
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      run: |
+        VERSION=v${{ steps.validate.outputs.version }}
+        gh pr create -B master -H release/${VERSION} --title "release ${VERSION}" --body "Release branch and changes created by GitHub Actions. This PR should include changes to the docs, CHANGELOG and local environment files."
```
Comment (JoelSpeed) [inline .github/workflows/create-release.yml:117]: Nit, missing new line character at the end here

File: .github/workflows/publish-release.yml [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+      with:
+        ref: ${{ github.event.pull_request.merge_commit_sha }}
+        fetch-depth: 0
+        fetch-tags: true
+
+    - name: Tag release
+      run: |
+        # Set up github-actions[bot] user
+        git config --local user.name "github-actions[bot]"
+        git config --local user.email "41898282+github-actions[bot]@users.noreply.github.com"
```
Comment (JoelSpeed) [inline .github/workflows/publish-release.yml:33]: Where did this magic bot name come from?

File: .github/workflows/publish-release.yml [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        password: ${{ secrets.REGISTRY_PASSWORD }}
+
+    - name: Build images
+      run: |
+        make docker-all
+    
+    - name: Push images
+      run: |
+        make docker-push-all
+
```
Comment (JoelSpeed) [inline .github/workflows/publish-release.yml:125]: Nit, whitespace on final line, should end with newline char

File: .github/workflows/publish-release.yml [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+      uses: docker/setup-qemu-action@v2
+    
+    - name: Set up Docker Buildx
+      id: buildx
+      uses: docker/setup-buildx-action@v2
+
+    - name: Login to quay.io
+      uses: docker/login-action@v2
+      with:
+        registry: quay.io/tuunit
```
Comment (JoelSpeed) [inline .github/workflows/publish-release.yml:114]: Wrong registry

File: Makefile [hunk 5]
```diff
@@ -41,7 +41,7 @@ $(BINARY):
 
 DOCKER_BUILD_PLATFORM         ?= linux/amd64,linux/arm64,linux/ppc64le,linux/arm/v6,linux/arm/v7
 DOCKER_BUILD_RUNTIME_IMAGE    ?= alpine:3.18
-DOCKER_BUILDX_ARGS            ?= --build-arg RUNTIME_IMAGE=${DOCKER_BUILD_RUNTIME_IMAGE}
+DOCKER_BUILDX_ARGS            ?= --build-arg RUNTIME_IMAGE=${DOCKER_BUILD_RUNTIME_IMAGE} --provenance=false
```
Comment (JoelSpeed) [inline Makefile:44]: What does this new arg do?

File: .github/workflows/publish-release.yml (hunk 2 above)
Comment (tuunit) [reply .github/workflows/publish-release.yml:33]: It is the "official" email address of the github actions bot. GitHub does not officially mention it in the documentation but you find it used everywhere. I guess people just looked at the commit history and figured out that's the email and user id github is using for it's bot. Even the official google release action is using this bot internally: https://github.com/google-github-actions/release-please-action/blob/main/.github/workflows/release-please.yaml#L76

File: Makefile (hunk 5 above)
Comment (tuunit) [reply Makefile:44]: It is used to control how much metadata is attached to images to figure out their origin:

https://docs.docker.com/build/attestations/
https://docs.docker.com/build/attestations/slsa-provenance/

I was trying to figure out why the automated build and the build from my machine produces these `unknown` archs for each regular arch that is build:
image

It seems like something is differently setup on your system as all previous builds to quay.io from your system do not contain any of these `unknown` archs.

Comment (tuunit) [reply Makefile:44]: I don't believe it is an issue but something is different between how I build the images and how you build images on your system.

You find the some "issue" for all nightly builds as well:
https://quay.io/repository/oauth2-proxy/oauth2-proxy-nightly?tab=tags

**Review by JoelSpeed — APPROVED** [review]

**Pull request merged**